e2e/test-runner/wsl_tester.go

```diff
@@ -0,0 +1,82 @@
+package test_runner
+
+import (
+	"flag"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"testing"
+)
+
+// Defaults choosen based on the default state of the repository on git-clone.
+const defaultLauncherName = "ubuntudev.launchername.dev.exe"
+const defaultDistroName = "UbuntuDev.WslId.Dev"
+
+const serverLogPath = "/var/log/installer/systemsetup-server-debug.log"
+const clientLogPath = "ubuntu-desktop-installer/packages/ubuntu_wsl_setup/build/windows/runner/Debug/.ubuntu_wsl_setup.exe/ubuntu_wsl_setup.exe.log"
+
+var launcherName = flag.String("launcher-name", defaultLauncherName, "WSL distro launcher under test.")
+var distroName = flag.String("distro-name", defaultDistroName, "WSL distro instance registered for testing.")
+
+// A testing wrapper with functions simplifying composing and running commands through os.Exec and checking their outputs.
+type Tester struct {
+	*testing.T
+}
+
+func checkTestEnvMinRequirements(t *testing.T) {
+	rootDir := os.Getenv("GITHUB_WORKSPACE")
+	if len(rootDir) == 0 {
+		t.Fatalf("GITHUB_WORKSPACE not set. It should point to the launcher repo root directory.")
+	}
+}
+
+// Returns an instance of Tester set for unregistering the distro instance on clean up.
+func WslTester(t *testing.T) Tester {
+	checkTestEnvMinRequirements(t)
+	tester := Tester{t}
+	tester.Cleanup(func() {
+		// print debug logs
+		if tester.Failed() {
+			tester.Log("\n\n=== Server Debug Log ====")
+			logContents := tester.AssertWslCommand("cat", serverLogPath)
+			tester.Logf("%s", logContents)
+			tester.Log("\n\n=== Client Debug Log ====")
+			rootDir := os.Getenv("GITHUB_WORKSPACE")
+			path := filepath.Join(rootDir, clientLogPath)
+			clientLogContents, err := ioutil.ReadFile(path)
+			if err != nil {
+				tester.Fatalf("%s", err)
+			}
+			tester.Logf("%s", clientLogContents)
+		}
+		// unregister the instance.
+		tester.AssertOsCommand("wsl.exe", "--unregister", *distroName)
+	})
+	return tester
+}
+
+// Runs 'wsl.exe -d distroName' command line supplied as arguments to the wsl.exe command.
+// Asserts that the process succeeds and returns its combined output (stdout + stderr).
+func (t *Tester) AssertWslCommand(args ...string) string {
+	fullArgs := append([]string{"-d", *distroName}, args...)
+	return t.AssertOsCommand("wsl.exe", fullArgs...)
+}
+
+// Runs the command line supplied as arguments to the configured distro launcher on powershell.
+// Asserts that the process succeeds and returns its combined output (stdout + stderr).
+func (t *Tester) AssertLauncherCommand(fullCommand string) string {

```

Comment:
Could be a good idea to allow for an extra argument for an error explanation:
```go
t.AssertLauncherCommand("test -f /var/log/upgrade-policy-changed.log", "Launcher did not crate sentinel file after changing upgrade policy.")
```
As tests get more abundant and specific we'll find these messages useful to quickly triage errors.

Not married to the idea, so please provide your thoughts.

Reply:
Well, the semantics of an error depends on which command is run. If a command exits 0 and produces no output when you expect for an output is in fact an error. So, to follow this idea we'd need to also pass a predicate function informing whether the command output means an error or not.

Still, the string can be supplied as a customization point for the cases when the command exits non-zero. Again, whether this is the only error situation possible or not depends on the command and expectations.

Reply:
Also, it hurts the API a little bit, because the current way allows passing a variadic number of strings composing the command line arguments. Allowing an error message to be passed require being clever on how to differentiate that string from the rest and not passing it to the command line. At a first glance it seems we'd need to add a type for that.